cli/npm/resolution.rs

```diff
@@ -415,32 +424,49 @@ impl NpmResolutionSnapshot {
           dependencies,
         };
 
-        packages.insert(package_id.clone(), package);
+        packages.insert(package_id, package);
+      }
+
+      // verify that all these ids exist in packages
+      for id in &verify_ids {
+        if !packages.contains_key(id) {
+          bail!(
+            "the lockfile ({}) is corrupt. Recreate it or run again with --no-lock",
+            lockfile.filename.display(),
+          );
+        }
       }
     }
 
     let mut unresolved_tasks = Vec::with_capacity(packages_by_name.len());
 
-    for package_id in packages.keys() {
-      let package_id = package_id.clone();
+    // cache the package names in parallel in the registry api
+    for package_name in packages_by_name.keys() {
+      let package_name = package_name.clone();
       let api = api.clone();
       unresolved_tasks.push(tokio::task::spawn(async move {
-        let info = api.package_info(&package_id.name).await?;
-        Result::<_, AnyError>::Ok((package_id, info))
+        api.package_info(&package_name).await?;
+        Result::<_, AnyError>::Ok(())
       }));
     }
     for result in futures::future::join_all(unresolved_tasks).await {
-      let (package_id, info) = result??;
-
-      let version_and_info = get_resolved_package_version_and_info(
-        &package_id.name,
-        &NpmVersionReq::parse(&package_id.version.to_string()).unwrap(),
-        info,
-        None,
-      )?;
+      result??;
+    }
 
-      let package = packages.get_mut(&package_id).unwrap();
-      package.dist = version_and_info.info.dist;
+    // ensure the dist is set for each package
+    for package in packages.values_mut() {
+      // this will read from the memory cache now
+      let package_info = api.package_info(&package.id.name).await?;
+      let version_info = match package_info
+        .versions
+        .get(&package.id.version.to_string())
+      {
+        Some(version_info) => version_info,
+        None => {
+          bail!("could not find '{}' specified in the lockfile. Maybe try again with --reload", package.id);
+        }
+      };
+      package.dist = version_info.dist.clone();
     }
```

Comment:
I had more code here initially, but deleted it because it would be a lot more to maintain and maybe not desirable. Let's look into it maybe after peer deps are added (as it creates some complications with that too).

```rs
      // ensure the dependencies align with the lockfile dependencies
      if package.dependencies.len() != version_info.dependencies.len() {
        bail!(
          "the lockfile ({}) had {} dependencies for {}, but the npm registry said there were {}",
          lockfile.lock().filename.display(),
          package.dependencies.len(),
          package.id,
          version_info.dependencies.len(),
        );
      }
      for entry in version_info.dependencies_as_entries()? {
        match package.dependencies.get(&entry.name) {
          Some(dep) => {
            if !entry.version_req.satisfies(&dep.version) {
              bail!(
                "the lockfile ({}) had a dependency requirement of {} for {}, but this didn't satisfy the npm registry's version requirement",
                lockfile.lock().filename.display(),
                // todo...
              );
            }
          }
          None => {
            // todo...better error
            bail!(
              "the lockfile ({}) is corrupt. Recreate it or run again with --no-lock",
              lockfile.lock().filename.display(),
            );
          }
        }
      }
```